src/sentry/migrations/0396_add_usecase_to_indexer.py

```diff
@@ -0,0 +1,57 @@
+# Generated by Django 2.2.28 on 2023-03-16 22:02
+
+from django.db import migrations, models
+
+from sentry.new_migrations.migrations import CheckedMigration
+
+
+class Migration(CheckedMigration):
+    # This flag is used to mark that a migration shouldn't be automatically run in production. For
+    # the most part, this should only be used for operations where it's safe to run the migration
+    # after your code has deployed. So this should not be used for most operations that alter the
+    # schema of a table.
+    # Here are some things that make sense to mark as dangerous:
+    # - Large data migrations. Typically we want these to be run manually by ops so that they can
+    #   be monitored and not block the deploy for a long period of time while they run.
+    # - Adding indexes to large tables. Since this can take a long time, we'd generally prefer to
+    #   have ops run this and not block the deploy. Note that while adding an index is a schema
+    #   change, it's completely safe to run the operation after the code has deployed.
+    is_dangerous = False
+
+    dependencies = [
+        ("sentry", "0395_add_index_to_date_added"),
+    ]
+
+    operations = [
+        migrations.SeparateDatabaseAndState(
+            database_operations=[
+                migrations.RunSQL(
+                    """
+                    ALTER TABLE "sentry_perfstringindexer" ADD COLUMN "use_case_id" varchar(120) NOT NULL DEFAULT 'performance';
+                    """,
+                    reverse_sql="""
+                    ALTER TABLE "sentry_perfstringindexer" DROP COLUMN "use_case_id";
+                    """,
+                    hints={"tables": ["sentry_perfstringindexer"]},
+                ),
+            ],
+            state_operations=[
+                migrations.AddField(
+                    model_name="perfstringindexer",
+                    name="use_case_id",
+                    field=models.CharField(default="performance", max_length=120),
+                ),
+                migrations.RemoveConstraint(
+                    model_name="perfstringindexer",
+                    name="perf_unique_org_string",
+                ),
+                migrations.AddConstraint(
+                    model_name="perfstringindexer",
+                    constraint=models.UniqueConstraint(
+                        fields=("string", "organization_id", "use_case_id"),
+                        name="perf_unique_org_string_usecase",
+                    ),
+                ),
```

Comment:
These changes don't have any DDL queries being generated for them. You could:

1. add the new index without validation
2. enable validation on the new index. Doing this separately reduces the duration of locks in postgres
3. then drop the old index.

The [develop docs](https://develop.sentry.dev/database-migrations/#adding-not-null-to-columns) have an example of the SQL used to add constraints as invalid and then validate them.

Reply:
This table has 1mm rows, I wonder how bad adding a unique index to a table like that would be.

My thought here is that maybe we just want ops to run the migration to create the index for us, instead of trying to do it as part of the deploy. We have a statement timeout when running migrations now, so if it takes longer than a few seconds it'll fail.

So my thought here is:
 - Keep the `AddField` in this migration, and let that run in the deploy
 - Make another migration, and move the `(Add|Remove)Constraint` operations there, and then mark is as `is_dangerous` so that ops can run manually.
     - I'd probably also recommend running the `AddConstraint` first, so that we're not running without a unique constraint at any point.

Reply:
@wedamija Your suggestion makes sense to me. @onewland do you have any thoughts?

Reply:
I'd defer to their team on the safe way to do this. Though thinking about it @wedamija would a unique index enforce the unique constraint indirectly? Because we're going to need that anyway

Reply:
Unique constraints in postgres are implemented via a unique index. The constraint is just a layer on top. You can actually just add a unique index and have no constraint and it'll work fine.

Reply:
Just for reference: https://www.postgresql.org/docs/current/indexes-unique.html#:~:text=PostgreSQL%20automatically%20creates%20a%20unique,mechanism%20that%20enforces%20the%20constraint.

> PostgreSQL automatically creates a unique index when a unique constraint or primary key is defined for a table. The index covers the columns that make up the primary key or unique constraint (a multicolumn index, if appropriate), and is the mechanism that enforces the constraint.

Reply:
Ok, so if I'm following, creating a separate migration that performs the `AddConstraint` operation would be sufficient for creating the unique index we want?

Reply:
Yep that'd be fine, and we'd just have ops run that for us to avoid blocking the deploy.